src/Registration/cReg/NiftyResampler.cpp

```diff
@@ -240,7 +240,7 @@ static void set_post_resample_outputs(std::shared_ptr<DataContainer> &out_to_ret
         NumberType::Type output_num_type = (*output_to_return_sptr->begin()).get_typeID();
         if (output_num_type != NumberType::CXFLOAT)
             throw std::runtime_error("NiftyResampler: Only complex type currently supported is complex float");
-        ImageData<dataType> ::Iterator &it_out = output_to_return_sptr->begin();
+        typename ImageData<dataType> ::Iterator &it_out = output_to_return_sptr->begin();
```

Comment:
much simpler to say
```suggestion
        auto &it_out = output_to_return_sptr->begin();
```
but are you sure it needs to be a reference? It's going to change, and therefore affect `output_to_return_sptr->begin()`, which seems terribly dangerous. (This is actually also true for `it_real` etc)

Reply:
I cannot be sure about anything here - this is Richard's code, I simply had to add `typename` for SIRF to build (now it builds but synergistic tests fail).

Is `auto` clever enough to use `typename`?

Reply:
virtual method's return types must be covariant, must they not?

Reply:
not sure what you mean. Virtual methods return types have indeed to be either identical or returning a derived version of something related to the base class (I haven't checked exact definition of "covariant return types", see maybe https://stackoverflow.com/questions/25813435/what-exactly-are-covariant-return-types-in-c).

However, that doesn't really apply here as far as I can see. I was merely suggestion to say
```c++
auto it_out = output_to_return_sptr->begin();
```
(if that compiles anyway)

Reply:
AFAIK covariant returns must be either pointers or references

Reply:
agreed. But as it's a reference, it might be possible to just say `auto`. Not sure...

Reply:
from the link you quoted: "Covariance only happens when you are returning a pointer or a reference to a class, and the classes are related by inheritance."